### PR TITLE
CSP-469 - Create Partner POST

### PR DESCRIPTION
**Pull request by LiamWright**
/Partners POST endpoint

Example request:

Header: `B46C2A2A-E35C-4788-B4B7-1F7E84081846`

Request:
```
{
  "userName": "userName",
  "email": "email@email.com",
  "name": "string",
  "joined": "2023-01-23T12:33:05.340Z",
  "regions": [
    1
  ],
  "information": "string",
  "organisation": "string"
}
```

### CODE REVIEW

File: src/SFA.DAS.AANHub.Database/Tables/Partner.sql [hunk 1]
```diff
@@ -1,10 +1,10 @@
 ﻿CREATE TABLE [dbo].[Partner]
 (
 	[MemberId] UNIQUEIDENTIFIER NOT NULL PRIMARY KEY, 
-    [UKPRN] BIGINT NOT NULL, 
-    [Email] NVARCHAR(256) NULL, 
-    [Name] NVARCHAR(200) NULL, 
-    [LastUpdated] DATETIME NULL, 
+    [Organisation] NVARCHAR(200) NOT NULL,
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Database/Tables/Partner.sql:4]: What is expected in the organisation field?

File: src/SFA.DAS.AANHub.Api/Models/CreatePartnerModel.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        public string UserName { get; set; } = null!;
+        public string Email { get; set; } = null!;
+        public string Name { get; set; } = null!;
+        public DateTime Joined { get; set; }
+        public List<int>? Regions { get; set; }
+        public string? Information { get; set; }
+        public string Organisation { get; set; } = null!;
+
+        public static implicit operator CreatePartnerMemberCommand(CreatePartnerModel model) => new()
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api/Models/CreatePartnerModel.cs:15]: Coverage missing for the operator

File: src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    public class CreatePartnerMemberCommand : CreateMemberCommandBase, IRequest<ValidatableResponse<CreatePartnerMemberCommandResponse>>, IRequestedByMemberId
+    {
+        public string UserName { get; set; } = null!;
+        public string Organisation { get; set; } = null!;
+        public Guid? RequestedByMemberId { get; set; }
+
+        public static implicit operator Member(CreatePartnerMemberCommand command) => new()
+        {
+            Id = command.Id,
+            UserType = MembershipUserType.Admin,
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs:19]: Is the Admin user type correct?

File: src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace SFA.DAS.AANHub.Application.Partners
+{
+    public class CreatePartnerMemberCommand : CreateMemberCommandBase, IRequest<ValidatableResponse<CreatePartnerMemberCommandResponse>>, IRequestedByMemberId
+    {
+        public string UserName { get; set; } = null!;
+        public string Organisation { get; set; } = null!;
+        public Guid? RequestedByMemberId { get; set; }
+
+        public static implicit operator Member(CreatePartnerMemberCommand command) => new()
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs:16]: Coverage missing for the operator?

File: src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommandValidator.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        public const string PartnerAlreadyExistsErrorMessage = "Username already exists";
+
+        public CreatePartnerMemberCommandValidator(IRegionsReadRepository regionsReadRepository, IMembersReadRepository membersReadRepository,
+            IPartnersReadRepository partnersReadRepository)
+        {
+            Include(new CreateMemberCommandBaseValidator(regionsReadRepository));
+            Include(new RequestedByMemberIdValidator(membersReadRepository));
+            RuleFor(c => c.UserName)
+                .NotEmpty()
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommandValidator.cs:18]: Validation for organisation is missing?

File: src/SFA.DAS.AANHub.Application.UnitTests/Partners/Commands/CreatePartnersMemberCommandValidatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+        public CreatePartnerMemberCommandValidatorTests()
+        {
+            _regionsReadRepository = new Mock<IRegionsReadRepository>();
+            _membersReadRepository = new Mock<IMembersReadRepository>();
+            _partnersReadRepository = new Mock<IPartnersReadRepository>();
+        }
+
+        [TestCase("userName", true)]
+        [TestCase("", false)]
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application.UnitTests/Partners/Commands/CreatePartnersMemberCommandValidatorTests.cs:23]: Perhaps add a test case for white space

File: src/SFA.DAS.AANHub.Database/Tables/Partner.sql (hunk 1 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Database/Tables/Partner.sql:4]: It's the employer name, I believe this same field is in the employer request as well

File: src/SFA.DAS.AANHub.Api/Models/CreatePartnerModel.cs (hunk 2 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Api/Models/CreatePartnerModel.cs:15]: It's covered as part of `CreatePartnersMemberCommandHandlerTests`, isn't it?

File: src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs (hunk 3 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs]: No, well spotted. This has been updated to `.Partner`

File: src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs (hunk 4 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs:16]: I think this is covered as part of `CreatePartnersMemberCommandHandlerTests`

File: src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommandValidator.cs (hunk 5 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommandValidator.cs:18]: This has now been added

File: src/SFA.DAS.AANHub.Api/Models/CreatePartnerModel.cs (hunk 2 above)
Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Api/Models/CreatePartnerModel.cs:15]: I expect to see explicit test, with assertions on each field.

File: src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs (hunk 4 above)
Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs:16]: Same here, explicit test on the operator is expected with assertion on each field. Especially here as we have applied default values.

File: src/SFA.DAS.AANHub.Api/Models/CreatePartnerModel.cs (hunk 2 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Api/Models/CreatePartnerModel.cs:15]: Added here now: https://github.com/SkillsFundingAgency/das-aan-hub-api/blob/7d8c7338247663f8460af149a795a511a4610dd1/src/SFA.DAS.AANHub.Api.UnitTests/Controllers/PartnersControllerTests.cs#L79

File: src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs (hunk 4 above)
Comment (LiamWright) [reply src/SFA.DAS.AANHub.Application/Partners/CreatePartnerMemberCommand.cs:16]: Added here: https://github.com/SkillsFundingAgency/das-aan-hub-api/blob/7d8c7338247663f8460af149a795a511a4610dd1/src/SFA.DAS.AANHub.Api.UnitTests/Controllers/PartnersControllerTests.cs#L79

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**